src/qinfer/geometry.py

```diff
@@ -0,0 +1,279 @@
+    # in our set of a convex combination, we need to construct the outer
+    # product
+    #     P_{ik} = sum_j y_ij y_jk
+    # where y_ij is the jth component of the ith point in our set.
+    P = cvxopt.matrix(np.dot(convex_points, convex_points.T))
+    # We have no linear objective, so we set q = 0.
+    # Similarly, q is a vector of all ones representing the sum
+    # over points in the set.
+    q = cvxopt.matrix(np.zeros((n_points, 1)))
+    # To enfoce x_i ≥ 0, we need G = -\mathbb{1} and h = 0,
+    # such that CVXOPT's constraint Gx ≥ h is correct.
+    G = cvxopt.matrix(-np.eye(n_points))
+    h = cvxopt.matrix(0.0, (n_points, 1))
+
+    # We do, however, have a linear constraint A x = 1.
+    A = cvxopt.matrix(1.0, (1, n_points))
+    b = cvxopt.matrix(1.0)
+
+    # Now we solve with cvxopt.
+    solution = cvxopt.solvers.qp(P, q, G, h, A, b,
+        options={'show_progress': False}
+    )
+
+    # Return the obtained primal objective, noting that CVXOPT finds
+    # min_x 1/2 x^T P X, so that we must multiply by 2, then take the
+    # square root. Taking the square root, we have to be a little
+    # careful, since finite tolerances can cause the objective to
+    # be negative, even though we constrained it to be at least zero.
+    # Thus, we take the max of the objective and zero.
+    return np.sqrt(2 * max(0, solution['primal objective']))
+
+def rand_argmin(arr, tol=1e-7):
+    minimum = np.min(arr)
+    return np.random.choice(np.nonzero(arr <= minimum + tol)[0])
+
+def expand_hull_directed_search(partial_hull, candidate_points, interior_thresh=1e-7):
+    r"""
+    Uses the Sartipizadeh and Vincent 2016 algorithm to expand
+    an approximate convex hull by finding a point :math:`\vec{x}`
+    solving
+
+    .. math::
+
+        \vec{x} = \operatorname{arg\,min}_{\vec{x} \in S \setminus E}
+                  \max_{\vec{z} \in S \setminus E}
+                  d(\vec{z}, E \cup \{\vec{x}\})
+
+    where :math:`E` is the set of points comprising the approximate
+    hull so far, and where :math:`S` is the set of points that we
+    wish to approximate the convex hull of. That is, this function
+    finds a point to add to an approximate hull that minimizes
+    the worst-case distance to any point not currently in the hull.
+    """
+    # TODO: document this much better, esp. params/returns.
+
+    # returns:
+    #    index of point to add to the hull
+    #    current error of the approximate hull
+    #    masked matrix of known convex distances
+
+    # Our strategy will be to build up a matrix D_{i, j}
+    # defined as:
+    #
+    #     Dᵢⱼ ≔ d(xᵢ, E ∪ {xⱼ})
+    #
+    # where the set of candidate points S \ E = {xᵢ} is represented
+    # as the array candidate_points.
+    #
+    # NB: our notation is different from VC16, since they reuse
+    # the letter E.
+
+    n_partial, dimension = partial_hull.shape
+    n_candidates, dimension_check = candidate_points.shape
+    assert dimension == dimension_check
```

Comment:
Absolutely. Given that this is still a first stab, I didn't yet take the care to validate inputs and all that will go into the final PR.